### ISSUE TITLE
HHEd fails during monophone MMF generation

### ISSUE THREAD

**Issue by kwojt42** [RHVoice/RHVoice#1053]
### 
HHEd fails during monophone MMF generation with "variance floor macro varFloor1 not found"
## Environment

**RHVoice commit**

```
e198383f8afd70c8afef4fa02c879ea6159dcaab
```

**Platform**

Ubuntu 24.04 (WSL2)

**Training tools**

HTS-HTK 2.3 (2015 uratec build)

```
HCompV
HInit
HRest
HERest
HHEd
HSMMAlign
HMGenS
```

RHVoice was initially configured with an old HTK installation from `/usr/local/bin`, but this was corrected and all HTS tools now point to the HTS-HTK 2.3 installation.

---

## Problem

Voice building proceeds successfully until the first HHEd stage ("making monophone MMF").

The pipeline successfully completes:

* import-recordings
* segmentation
* label generation
* feature extraction
* cmp generation
* question generation
* LPF generation
* variance floor computation
* HInit
* HRest

Training stops only in HHEd.

The reported error is:

```
ERROR [+7023] ApplyVFLoor:
variance floor macro varFloor1 not found
```

---

## Files verified

The following files were checked and generated correctly:

```
labels/
cmp/
lists/
scp/
models/qst001/ver1/cmp/vFloors
edfiles/qst001/ver1/cmp/lvf.hed
```

`train.scp`, `mono.list` and `full.list` are also generated correctly.

---

## HHEd command executed

Training.pl executes:

```text
/home/ristantal/tts/hts23/bin/HHEd \
-r 1 \
-A -B \
-C configs/qst001/ver1/trn.cnf \
-D -T 1 \
-p -i \
-d models/qst001/ver1/cmp/HRest \
-w models/qst001/ver1/cmp/monophone.mmf \
edfiles/qst001/ver1/cmp/lvf.hed \
data/lists/mono.list
```

---

## Why this looks suspicious

The generated `lvf.hed` contains:

```
FV ".../models/qst001/ver1/cmp/vFloors"

TI SW_all {*.state[2-6].weights}
```

The generated `vFloors` contains all expected variance macros:

```
~v "varFloor1"
~v "varFloor2"
~v "varFloor3"
~v "varFloor4"
~v "varFloor5"
```

During execution HHEd reports:

```
loading vector 1
loading vector 2
loading vector 3
loading vector 4
loading vector 5
```

Immediately afterwards it reports:

```
ApplyVFLoor:
variance floor macro varFloor1 not found
```

It appears that HHEd successfully loads the variance floor file but later cannot resolve the loaded macro.

---

## Investigation already performed

The following were verified:

* labels generated correctly
* cmp files generated correctly
* train.scp generated correctly
* mono.list generated correctly
* full.list generated correctly
* HInit completed successfully
* HRest completed successfully
* vFloors generated correctly
* lvf.hed generated correctly
* HTS binaries verified
* exact HHEd command printed from Training.pl

At this point I am no longer able to determine whether this is caused by:

* an incompatibility between the current RHVoice training scripts and HTS-HTK 2.3,
* a regression in the training scripts,
* or an issue inside the HHEd stage.

Any advice would be greatly appreciated.

Attached:

* log
* training-debug.log
* vFloors
* lvf.hed

HHEd fails during monophone MMF generation with "variance floor macro varFloor1 not found"
## Environment

**RHVoice commit**

```
e198383f8afd70c8afef4fa02c879ea6159dcaab
```

**Platform**

Ubuntu 24.04 (WSL2)

**Training tools**

HTS-HTK 2.3 (2015 uratec build)

```
HCompV
HInit
HRest
HERest
HHEd
HSMMAlign
HMGenS
```

RHVoice was initially configured with an old HTK installation from `/usr/local/bin`, but this was corrected and all HTS tools now point to the HTS-HTK 2.3 installation.

---

## Problem

Voice building proceeds successfully until the first HHEd stage ("making monophone MMF").

The pipeline successfully completes:

* import-recordings
* segmentation
* label generation
* feature extraction
* cmp generation
* question generation
* LPF generation
* variance floor computation
* HInit
* HRest

Training stops only in HHEd.

The reported error is:

```
ERROR [+7023] ApplyVFLoor:
variance floor macro varFloor1 not found
```

---

## Files verified

The following files were checked and generated correctly:

```
labels/
cmp/
lists/
scp/
models/qst001/ver1/cmp/vFloors
edfiles/qst001/ver1/cmp/lvf.hed
```

`train.scp`, `mono.list` and `full.list` are also generated correctly.

---

## HHEd command executed

Training.pl executes:

```text
/home/ristantal/tts/hts23/bin/HHEd \
-r 1 \
-A -B \
-C configs/qst001/ver1/trn.cnf \
-D -T 1 \
-p -i \
-d models/qst001/ver1/cmp/HRest \
-w models/qst001/ver1/cmp/monophone.mmf \
edfiles/qst001/ver1/cmp/lvf.hed \
data/lists/mono.list
```

---

## Why this looks suspicious

The generated `lvf.hed` contains:

```
FV ".../models/qst001/ver1/cmp/vFloors"

TI SW_all {*.state[2-6].weights}
```

The generated `vFloors` contains all expected variance macros:

```
~v "varFloor1"
~v "varFloor2"
~v "varFloor3"
~v "varFloor4"
~v "varFloor5"
```

During execution HHEd reports:

```
loading vector 1
loading vector 2
loading vector 3
loading vector 4
loading vector 5
```

Immediately afterwards it reports:

```
ApplyVFLoor:
variance floor macro varFloor1 not found
```

It appears that HHEd successfully loads the variance floor file but later cannot resolve the loaded macro.

---

## Investigation already performed

The following were verified:

* labels generated correctly
* cmp files generated correctly
* train.scp generated correctly
* mono.list generated correctly
* full.list generated correctly
* HInit completed successfully
* HRest completed successfully
* vFloors generated correctly
* lvf.hed generated correctly
* HTS binaries verified
* exact HHEd command printed from Training.pl

At this point I am no longer able to determine whether this is caused by:

* an incompatibility between the current RHVoice training scripts and HTS-HTK 2.3,
* a regression in the training scripts,
* or an issue inside the HHEd stage.

Any advice would be greatly appreciated.

Attached:

* log
* training-debug.log
* vFloors
* lvf.hed

**Comment by zstanecic** [comment]
This is the problem I found also while using new ubuntu. I recommend sticking with debian when it comes to training with wsl.

**Comment by rmcpantoja** [comment]
Hi.

You meant uratec build. Is that different than the official builds?

Newer hts/htk archives works for newer ubuntu. Tested and confirmed.

The website for these tools went down, but I have a backup resting here some years ago.

Cheers.

**Comment by zstanecic** [comment]
No, new archives doesn't work on newer ubuntu 24.04. 
I have called many times to the cambridge university via the phone regarding the htk. Maybe it is a time still to write the email.
Their website is falling from time to time.

**Comment by zstanecic** [comment]
@rmcpantoja btw, htk site is working again. So, I encourage to abide by the license terms here.

**Comment by kwojt42** [comment]
Thank you both for your replies.

The HTS installation itself appears to be working correctly after switching to the HTS-HTK 2.3 (uratec) build.

The issue I am trying to understand is different.

The complete training pipeline succeeds up to the first HHEd stage:

* labels
* cmp generation
* variance floor computation
* HInit
* HRest

The failure occurs only during the first HHEd invocation.

The generated `vFloors` file exists and HHEd reports:

```
loading vector 1
loading vector 2
loading vector 3
loading vector 4
loading vector 5
```

Immediately afterwards it aborts with:

```
ApplyVFLoor:
variance floor macro varFloor1 not found
```

This is the part I cannot explain.

Does anyone know whether this is a known incompatibility between the current RHVoice training scripts and HTS-HTK 2.3, or should I continue looking for a configuration problem in my training setup?

Any suggestions would be greatly appreciated.To bierzmy się za twoją propozycje sprawdzenia

**Comment by zstanecic** [comment]
What is uratec build? I don't understand your conclusions. Don't waste our time and use debian to train the voices. If one part of training is not working, it is not working at all.

**Comment by zstanecic** [comment]
Update: on ubuntu, compile all tools with older compiler (gcc)
